integrations/google/forms/events.go

```diff
@@ -0,0 +1,79 @@
+package forms
+
+import (
+	"context"
+	"strings"
+
+	"go.uber.org/zap"
+	"google.golang.org/api/forms/v1"
+
+	"go.autokitteh.dev/autokitteh/integrations/internal/extrazap"
+	"go.autokitteh.dev/autokitteh/internal/kittehs"
+	"go.autokitteh.dev/autokitteh/sdk/sdkservices"
+	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
+)
+
+func ConstructEvent(ctx context.Context, vars sdkservices.Vars, formsEvent map[string]any, cids []sdktypes.ConnectionID) (sdktypes.Event, error) {
+	l := extrazap.ExtractLoggerFromContext(ctx)
+
+	// Enrich the event with relevant data, using API calls.
+	if len(cids) > 0 {
+		a := api{vars: vars, cid: cids[0]}
+		switch WatchEventType(formsEvent["event_type"].(string)) {
+		case WatchSchemaChanges:
+			form, err := a.getForm(ctx)
+			if err != nil {
+				l.Error("Failed to get form", zap.Error(err))
+				// Don't abort, dispatch the event without this data.
+			}
+			formsEvent["form"] = form
+
+		case WatchNewResponses:
+			responses, err := a.listResponses(ctx)
+			if err != nil {
+				l.Error("Failed to list responses", zap.Error(err))
+				// Don't abort, dispatch the event without this data.
+			}
+			formsEvent["response"] = lastResponse(responses)
+		}
+	}
+
+	// Convert the raw data to an AutoKitteh event.
+	wrapped, err := sdktypes.DefaultValueWrapper.Wrap(formsEvent)
+	if err != nil {
+		l.Error("Failed to wrap Google Forms event", zap.Error(err))
```

Comment:
I think we should either log or return an error. If we return an error the caller need to log it if needed. wdyt?
See https://dave.cheney.net/tag/logging

Reply:
I'm assuming you're referring to [Let’s talk about logging](https://dave.cheney.net/2015/11/05/lets-talk-about-logging)? I disagree with many of the recommendations there. They do not apply to modern software systems, even when they were written, a decade ago. :)

In this function I think it's more useful to log the specific problem that would help a developer contextualize the function's failure during debugging, instead of burying it in the stack trace of a wrapped error object that isn't visible without double-clicking on the log entry.

To be clear, this is a case-by-case decision. Other functions may not benefit from this pin-pointing, or it may be too granular during debugging - in which case a single error message - as you suggest - is more appropriate.

Reply:
I agree with @daabr . I also support logging the problem where it happened instead/in addition to the caller